### PR TITLE
Rreverser html comments

### PR DESCRIPTION
**Pull request by tdewolff**
Amendment of https://github.com/tdewolff/parse/pull/24

### CODE REVIEW

File: js/lex.go [hunk 1]
```diff
@@ -289,35 +288,35 @@ func (l *Lexer) consumeUnicodeEscape() bool {
 	return true
 }
 
-////////////////////////////////////////////////////////////////
-
-func (l *Lexer) consumeSingleLineToken(skip int) {
-	l.r.Move(skip)
```
Comment (RReverser) [inline js/lex.go]: Curious - why did you move this out? Seems like a nice abstraction given that you always need to skip some prefix when consuming a line comment.

File: js/lex.go [hunk 2]
```diff
@@ -289,35 +288,35 @@ func (l *Lexer) consumeUnicodeEscape() bool {
...
 		} else if c >= 0xC0 {
-			mark := l.r.Pos()
+			//mark := l.r.Pos()
 			if r, _ := l.r.PeekRune(0); r == '\u2028' || r == '\u2029' {
-				l.r.Rewind(mark)
+				//l.r.Rewind(mark)
 				break
 			}
 		}
 		l.r.Move(1)
```
Comment (RReverser) [inline js/lex.go]: Unrelated to the changes, but I wonder how this will work for runes that are not end-of-line. Shouldn't we skip entire rune instead of a single byte in order not to appear in the middle of unicode character?

File: js/lex.go [hunk 3]
```diff
@@ -323,6 +321,12 @@ func (l *Lexer) consumeCommentToken() bool {
 				if c == '*' && l.r.Peek(1) == '/' {
 					l.r.Move(2)
 					return true
+				} else if c == '\r' || c == '\n' {
```
Comment (RReverser) [inline js/lex.go]: Maybe just `l.consumeLineTerminator()` like inside of strings etc.?

File: js/lex.go (hunk 1 above)
Comment (tdewolff) [inline js/lex.go]: Because at the call site the `skip` integer its purpose is unclear. Besides, no other consume function uses this pattern so it is not consistent. Every consume function assumes it is at the beginning of what must be consumed, passing `skip` violates that.

File: js/lex.go (hunk 2 above)
Comment (tdewolff) [inline js/lex.go]: Any UTF-8 bytes can never be mistaken for ASCII characters, see https://research.swtch.com/utf8

File: js/lex.go (hunk 3 above)
Comment (tdewolff) [inline js/lex.go]: Yes that would be better I think

File: js/lex.go (hunk 2 above)
Comment (RReverser) [inline js/lex.go]: Sure, it's more like optimization - if we anyway peeked the rune up, we might as well use the already-known length to skip it entirely instead of continuing the loop inside of it.

File: js/lex.go (hunk 1 above)
Comment (RReverser) [inline js/lex.go]: Ok, fine with me.

File: js/lex.go (hunk 2 above)
Comment (RReverser) [inline js/lex.go]: Basically, instead of peeking the rune, we can just consume it, and then if it was a newline, break the loop.

Comment (RReverser) [inline js/lex.go]: Basically, instead of peeking the rune, we can just consume it, and then if it was a newline, set the `emptyLine`.

**Comment by RReverser** [comment]
Few comments (about line terminators and skipping UTF-8), but otherwise changes LGTM, thanks.

File: js/lex.go (hunk 2 above)
Comment (tdewolff) [inline js/lex.go]: Hm yes, this might slightly improve performance. Usually UTF-8 codepoints are very rare (compared to ASCII), so that any execution paths for UTF-8 might be in the way for ASCII (as in CPU cache lines being missed). So if there is an improvement it would still be very small. Usually these performance predictions are wrong because it is so non-trivial ;-)

I will run a benchmark to prove it

Comment (RReverser) [inline js/lex.go]: Yeah, I don't think the difference will be significant, just that it's quite an easy things to do anyway, and might be easier to read / debug :)

Comment (RReverser) [inline js/lex.go]: Also concerns about cache line are already valid for rune peeking, and when we did it, cache already contains the memory with that rune.

**Comment by RReverser** [comment]
"Skip rune" nitpick aside, seems good to go now.

**Comment by RReverser** [comment]
(hope it's not rude to "LGTM" your changes given that it's your repo 😄  )

**Comment by tdewolff** [comment]
@RReverser Not at all! The skip-rune thing is included in the consumeLineTerminator function, no?

**Comment by RReverser** [comment]
@tdewolff I don't think so, `consumeLineTerminator` was in `consumeCommentToken` (multiline token part), while skip-rune suggestion was in `consumeSingleLineComment`.

**Comment by tdewolff** [comment]
Right, but `consumeSingleLineComment` doesn't consume the newlines

**Comment by RReverser** [comment]
Yes, the suggestion we discussed above was explicitly related to skipping runes when we already 1) peeked them up and 2) checked that they're not newlines. Doesn't matter too much though.

**Comment by tdewolff** [comment]
Ah yes now I remember. I will look into this separately, because then all occurences would need to be replaced.

**Comment by RReverser** [comment]
Cool, np.
